src/Coordination/Changelog.cpp

```diff
@@ -68,81 +77,293 @@ Checksum computeRecordChecksum(const ChangelogRecord & record)
 
-    void flush(bool force_fsync)
-    {
         if (compressed_buffer)
         {
-            /// Flush compressed data to WriteBufferFromFile working_buffer
+            /// Flush compressed data to file buffer
             compressed_buffer->next();
         }
 
-        WriteBuffer * working_buf = compressed_buffer ? compressed_buffer->getNestedBuffer() : file_buf.get();
+        last_index_written = record.header.index;
 
-        /// Flush working buffer to file system
-        working_buf->next();
+        return true;
+    }
 
+    void flush()
+    {
+        auto * file_buffer = tryGetFileBuffer();
         /// Fsync file system if needed
-        if (force_fsync)
-            working_buf->sync();
+        if (file_buffer && log_file_settings.force_sync)
+            file_buffer->sync();
     }
 
     uint64_t getStartIndex() const
     {
-        return start_index;
+        assert(current_file_description);
+        return current_file_description->from_log_index;
+    }
+
+    void rotate(uint64_t new_start_log_index)
+    {
+        /// Start new one
+        auto new_description = std::make_shared<ChangelogFileDescription>();
+        new_description->prefix = DEFAULT_PREFIX;
+        new_description->from_log_index = new_start_log_index;
+        new_description->to_log_index = new_start_log_index + log_file_settings.rotate_interval - 1;
+        new_description->extension = "bin";
+
+        if (log_file_settings.compress_logs)
+            new_description->extension += "." + toContentEncodingName(CompressionMethod::Zstd);
+
+        new_description->path = formatChangelogPath(
+            changelogs_dir,
+            new_description->prefix,
+            new_start_log_index,
+            new_start_log_index + log_file_settings.rotate_interval - 1,
+            new_description->extension);
+
+        LOG_TRACE(log, "Starting new changelog {}", new_description->path);
+        auto [it, inserted] = existing_changelogs.insert(std::make_pair(new_start_log_index, std::move(new_description)));
+
+        setFile(it->second, WriteMode::Rewrite);
+    }
+
+    void finalize()
+    {
+        if (isFileSet() && prealloc_done)
+            finalizeCurrentFile();
     }
 
 private:
+
+    void finalizeCurrentFile()
+    {
+        const auto * file_buffer = tryGetFileBuffer();
+        assert(file_buffer && prealloc_done);
+
+        assert(current_file_description);
+        // compact can delete the file and we don't need to do anything
+        if (current_file_description->deleted)
+        {
+            LOG_WARNING(log, "Log {} is already deleted", file_buffer->getFileName());
+            return;
+        }
+
+        if (log_file_settings.compress_logs)
+            compressed_buffer->finalize();
+
+        flush();
+
+        if (log_file_settings.max_size != 0)
+            ftruncate(file_buffer->getFD(), initial_file_size + file_buffer->count());
```

Comment:
I'm not sure that `count()` is reliable enough, it's too easy to calculate total bytes count incorrectly, especially for complicated buffers and for chains of buffers. AFAIK, `count()` should work fine for most write buffers, but, for example, it's just impossible to implement `count()` correctly for `PeaakableReadBuffer`

Reply:
I use `count()` for the `FileBuffer` so I would say that it's okay (don't even use it with ZSTD buffer).
Even if we have chained or more complex buffers we would always care here only how much is written to the last (file) buffer.
I will add a comment just in case.